test/cpp/end2end/async_end2end_test.cc

```diff
@@ -807,7 +807,11 @@ class AsyncEnd2endServerTryCancelTest : public AsyncEnd2endTest {
     EXPECT_FALSE(context->IsCancelled());
     context->TryCancel();
     gpr_log(GPR_INFO, "Server called TryCancel()");
-    EXPECT_TRUE(context->IsCancelled());
+    while (!context->IsCancelled()) {
```

Comment:
Agreed that the previous version was wrong, but this seems to remove the "async"ness of the test. The test will get more complicated, but I'm thinking that it needs a real state-based solution here.